admin-dev/themes/new-theme/js/components/grid/extension/preview-extension.js

```diff
@@ -0,0 +1,196 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+const $ = window.$;
+
+/**
+ * Extends grid with preview functionality.
+ */
+export default class PreviewExtension {
+  constructor(previewRenderer) {
+    this.lock = [];
+    this.renderer = previewRenderer;
+    this.expandSelector = '.js-expand';
+    this.collapseSelector = '.js-collapse';
+    this.previewOpenClass = 'preview-open';
+    this.previewToggleSelector = '.preview-toggle';
+  }
+
+  /**
+   * Extends provided grid with preview functionality
+   *
+   * @param grid
+   */
+  extend(grid) {
+    const $gridContainer = $(grid.getContainer);
+
+    $gridContainer.find('tbody tr').on('mouseover mouseleave', event => this._handleIconHovering(event));
+    $gridContainer.find(this.previewToggleSelector).on('click', event => this._togglePreview(event));
+  }
+
+  /**
+   * Shows/hides preview toggling icons
+   *
+   * @param event
+   * @private
+   */
+  _handleIconHovering(event) {
+    const $previewToggle = $(event.currentTarget).find(this.previewToggleSelector);
+
+    if (event.type === 'mouseover' && !$(event.currentTarget).hasClass(this.previewOpenClass)) {
+      this._showExpandIcon($previewToggle);
+    } else {
+      this._hideExpandIcon($previewToggle);
+    }
+  }
+
+  /**
+   * Shows/hides preview
+   *
+   * @param event
+   * @private
+   */
+  _togglePreview(event) {
+    const $previewToggle = $(event.currentTarget);
+    const $columnRow = $previewToggle.closest('tr');
+
+    if ($columnRow.hasClass(this.previewOpenClass)) {
+      $columnRow.next('.preview-row').remove();
+      $columnRow.removeClass(this.previewOpenClass);
+      this._showExpandIcon($columnRow);
+      this._hideCollapseIcon($columnRow);
+
+      return;
+    }
+    const dataUrl = $(event.currentTarget).data('preview-data-url');
+
+    if (this._isLocked(dataUrl)) {
+      return;
+    }
+
+    // Prevents loading preview multiple times.
+    // Uses "dataUrl" as lock key.
+    this._lock(dataUrl);
+
+    $.ajax({
+      url: dataUrl,
+      method: 'GET',
+      dataType: 'json',
+    }).then((response) => {
+      this._renderPreviewContent($columnRow, response.preview);
+
+      this._unlock(dataUrl);
+    }).catch((e) => {
+      showErrorMessage(e.responseJSON.message);
+
+      this._unlock(dataUrl);
+    });
+  }
+
+  /**
+   * Renders preview content
+   *
+   * @param $columnRow
+   * @param content
+   *
+   * @private
+   */
+  _renderPreviewContent($columnRow, content) {
+    const rowColumnCount = $columnRow.find('td').length;
+
+    const previewTemplate = `
```

Comment:
HTML in JS... are you a [JSX](https://www.w3schools.com/react/react_jsx.asp) fan ? :trollface: 

I guess here it's hard to not use it, but I dont feel good about it 😐 
Any suggestions @eternoendless @PierreRambaud ?

Reply:
I kind of like JSX. :D But I'm open for suggestions if you have any. :)

Reply:
I'm not a big fan to have this kind of template in JS, but, afaik there is no proper solution by using jquery / vanilla js (`<template>` is not supported everywhere). If only we were on VueJs here :sob:

Reply:
> I'm not a big fan to have this kind of template in JS, but, afaik there is no proper solution by using jquery / vanilla js (`<template>` is not supported everywhere). If only we were on VueJs here

web components maybe? With polyfills it also supports IE11 I think.

Reply:
Otherwise if you want to do it more simple, pass this template as PreviewExtension's parameter, and this template is stored in a twig template :thinking: 
At least, people will be able to improve/edit it.

Reply:
We would end up with something like this:
```twig
{% block preview_template %}
<tr class="preview-row">
      <td></td> {# Maybe even without <td> #}
</tr>
{% endblock %}
```

It would be so easy to break things up (we need to remember to add copy/paste template every time, it's easy to mess it up), developers are free implement their own previews if they wish to.

Does it really makes sense to move it away from JS and add dependency?